designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,230 @@
+
+#. Run through CI/CD pipeline for the application before publishing.
+
+#. Publish to other repositories besides SAR.
+
+#. Recognize template changes and suggest version number.
+
+#. Parse metadata from template without an AWS::ServerlessRepo::Application section.
+
+.. _application permission: https://docs.aws.amazon.com/serverlessrepo/latest/devguide/access-control-resource-based.html#application-permissions
+
+
+User Experience Walkthrough
+---------------------------
+
+Assuming that customers have the following SAM template:
+
+.. code-block:: yaml
+
+    Metadata:
+        AWS::ServerlessRepo::Application:
+            Name: my-app
+            Description: hello world
+            Author: user1
+            SpdxLicenseId: Apache-2.0
+            LicenseUrl: ./LICENSE.txt
+            ReadmeUrl: ./README.md
+            Labels: ['tests']
+            HomepageUrl: https://github.com/user1/my-app-project
+            SemanticVersion: 1.0.0
+            SourceCodeUrl: https://github.com/user1/my-app-project
+
+    Resources:
+        HelloWorldFunction:
+            Type: AWS::Lambda::Function
+            Properties:
+                CodeUri: s3://bucket/hello-world
+
+Package SAM template
+  Run ``sam package --template-file ./template.yaml --output-template-file packaged.yaml --s3-bucket my-bucket``
+  to upload code artifacts, readme and license files to S3 and generate the packaged template.
+
+Create new application in SAR
+  Run ``sam publish app -t ./packaged.yaml`` to publish a new application named my-app in SAR with the first version
+  created as 1.0.0. The app will be created as private by default. SAM CLI prints application created message and
+  link to the console details page.
+
+Create new version of an existing SAR application
+  Modify the existing template, give a different SemanticVersion value, and run ``sam publish app -t ./packaged.yaml``.
+  SAM CLI prints application metadata updated message, application version created message, values of the current application
+  metadata and link to the console details page.
+
+Crete application/version and set application permission
+  Run ``sam publish app -t ./packaged.yaml --make-public`` to publish the app and share it publicly. If ``--make-private``
+  option is used, the app will only be visible to the owner. If ``--account-ids <account ids>`` is used, the app will be
+  shared with the provided AWS accounts.
+
+Update the metadata of an exsiting application without creating new version
+  Keep SemanticVersion unchanged, then modify metadata fields like Description or ReadmeUrl, and run
+  ``sam publish app -t ./packaged.yaml``. SAM CLI prints application metadata updated message, values of the current
+  application metadata and link to the console details page.
+
+Output of the ``sam publish app`` command will be a link to the AWS Serverless Application Repository console details page
```

Comment:
Something I have been thinking about recently, is communicating next things customers can do after the command runs. You can use `sam build` as a guide/model for this.

Reply:
Nice suggestion, added what customers can do after the application is published.